karabo_data/run_files_map.py

```diff
@@ -0,0 +1,108 @@
+import json
+import logging
+import os
+import os.path as osp
+import re
+
+from .read_machinery import DATA_ROOT_DIR
+
+log = logging.getLogger(__name__)
+
+class RunFilesMap:
+    """Cached data about HDF5 files in a run directory
+
+    Stores the train IDs and source names in each file, along with some
+    metadata to check that the cache is still valid. The cached information
+    can be stored in:
+
+    - (run dir)/karabo_data_map.json
```

Comment:
For `raw` directory, no-one has write access. Do you know how it is for the `proc`? Do users can write to this directory or is it only for facility services?

Reply:
Users can't write to `proc` as far as I know.

I've included this location in the hope that if we've got the code, we can persuade ITDM to create the map file automatically when each run is written. Then no-one needs to wait several minutes to see what's in a run.

Reply:
OK, that makes sens :+1:

Reply:
I am optimistic this could be run as a service at the end of recording a run (eventually).

[We should also push for karabo-data-validate to run automatically.]

I see the attraction of JSON files: faster reading and plain text. I also see the attraction of sticking to one file format. This is a tough call.

To run lsxfel, is this Karabo_data_map the only file that needs to be read? If so, then I guess you can 'feel' the difference between a 400ms and a 40ms read time?

Reply:
If you `lsxfel` a run directory with detector data (AGIPD/LPD/DSSC), at present it will still open a single HDF5 file from that run to get some statistics like module size and frames per train. I'm thinking of adding this into the map file too, so it only needs to read that.

With this PR plus #207, I got timings down to about 0.3s for `lsxfel` on a large run. Using HDF5 for the cache would presumably make that about 0.6s. There's a rule of thumb that things happening in 0.1s feel instant to the user, so the difference would be noticeable.

There are probably ways to make an HDF5 map file that can be read faster - combine data into larger datasets so we can do fewer, bigger reads. But I've not come up with any way to do that and preserve convenient inspectability.